### PR TITLE
This is to make flume work in a kerberised cluster and make syslogs and chef-logs world readable.

### PR DESCRIPTION
**Pull request by saurabh313**
Added new keytab for flume so it can authenticate to the kerberised cluster and aggregate logs from all nodes and put in hdfs with public permission. 

Also, make flume roll over flume logs so it doesn't fill up local diskspace.

### CODE REVIEW

**Comment by saurabh313** [comment]
ubuntu@bcpc-vm1:~$ hdfs dfs -ls /user/flume/logs                                       
Found 2 items                                                                          
drwxr-xr-x   - flume supergroup          0 2016-03-24 14:47 /user/flume/logs/16-03-23  
drwxr-xr-x   - flume supergroup          0 2016-03-24 15:59 /user/flume/logs/16-03-24

**Comment by amithkanand** [comment]
Can you please squash the commits?

File: cookbooks/bcpc-hadoop/templates/default/flume_flume-conf.erb [hunk 1]
```diff
@@ -16,3 +16,5 @@
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollInterval = <%= node['bcpc']['hadoop']['copylog_rollup_interval'] %>
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollSize = 0
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollCount = 0
+<%= @agent_name %>.sinks.hdfsSink.hdfs.kerberosPrincipal = <%= node['bcpc']['hadoop']['kerberos']['data']['flume']['principal'] %>/<%= float_host(node[:fqdn]) %>@<%= node['bcpc']['hadoop']['kerberos']['realm'] %>
```
Comment (amithkanand) [inline cookbooks/bcpc-hadoop/templates/default/flume_flume-conf.erb]: This should include the logic to specifiy principal if a particular host is specified.

Comment (http-418) [inline cookbooks/bcpc-hadoop/templates/default/flume_flume-conf.erb]: This should also handle the possibility that flume is configured but kerberos is not.

**Comment by cbaenziger** [comment]
Saurabh could you name the log4j.conf to `flume_log4j.properties` to be consistent with the naming convention used in templates?

File: cookbooks/bcpc-hadoop/attributes/default.rb [hunk 2]
```diff
@@ -127,3 +127,9 @@
 default[:bcpc][:hadoop][:hdfs][:ldap][:password] =  nil
 default[:bcpc][:hadoop][:hdfs][:ldap][:search][:filter][:user]="(&(objectclass=user)(sAMAccountName={0}))"
 default[:bcpc][:hadoop][:hdfs][:ldap][:search][:filter][:group]="(objectClass=group)"
+#
+# Some log4j defaults
+#
+
+default[:bcpc][:hadoop][:log4j][:max_file_size] = "100MB"
+default[:bcpc][:hadoop][:log4j][:max_backups] = 10
```
Comment (cbaenziger) [inline cookbooks/bcpc-hadoop/attributes/default.rb]: These should be used in:
* [hb_log4.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/hb_log4j.properties.erb)
* [hdp_log4j.properties](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/hdp_log4j.properties.erb)
* [httpfs-log4j.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/httpfs-log4j.properties.erb)
* [hue_log4j.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/hue_log4j.properties.erb)
* [hv_hive-exec-log4j.properties](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/hv_hive-exec-log4j.properties.erb)
* [hv_hive-log4j.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/hv_hive-log4j.properties.erb)
* [ooz_oozie-log4j.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/ooz_oozie-log4j.properties.erb)
* [zk_log4j.properties.erb](https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/templates/default/zk_log4j.properties.erb)

File: cookbooks/bcpc-hadoop/recipes/copylog.rb [hunk 3]
```diff
@@ -49,11 +49,21 @@
...
   service "flume-agent-multi" do
     supports :status => true, :restart => true, :reload => false
     action [:enable, :start]
     subscribes :restart, "template[/etc/flume/conf/flume-env.sh]", :delayed
   end
+
+  template '/etc/flume/conf/log4j.properties' do
+    source "log4j.erb"
+    mode '0644'
+    owner 'root'
```
Comment (cbaenziger) [inline cookbooks/bcpc-hadoop/recipes/copylog.rb]: Probably 'flume:flume' instead of 'root:root'?

File: cookbooks/bcpc-hadoop/recipes/default.rb [hunk 4]
```diff
@@ -31,6 +31,11 @@
     'docopy' => false
 }
 
+node.default['bcpc']['hadoop']['copylog']['cheflog'] = {
+    'logfile' => "/var/log/chef/client.log",
```
Comment (cbaenziger) [inline cookbooks/bcpc-hadoop/recipes/default.rb]: This should be an attribute of the Chef-Client cookbook [file name](https://github.com/chef-cookbooks/chef-client/blob/2d09db1fd44f1be536034311dd853284133e1e75/attributes/default.rb#L38) and [file path](https://github.com/chef-cookbooks/chef-client/blob/2d09db1fd44f1be536034311dd853284133e1e75/attributes/default.rb#L46).

**Comment by saurabh313** [comment]
This should also solve the issue of [filling up root disk] (https://github.com/bloomberg/chef-bach/issues/168)

File: cookbooks/bcpc-hadoop/templates/default/flume_flume-conf.erb [hunk 5]
```diff
@@ -16,3 +16,7 @@
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollInterval = <%= node['bcpc']['hadoop']['copylog_rollup_interval'] %>
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollSize = 0
 <%= @agent_name %>.sinks.hdfsSink.hdfs.rollCount = 0
+<%= if node[:bcpc][:hadoop][:kerberos][:enable] %>
```
Comment (amithkanand) [inline cookbooks/bcpc-hadoop/templates/default/flume_flume-conf.erb]: This should be `<% if node[:bcpc][:hadoop][:kerberos][:enable] %>`

**Comment by amithkanand** [comment]
This doesn't seem to work on my test cluster. I cherry-picked the commit and cluster chef'd successfully. I can see the `flume-agent-mutli` service running for each kind of agent but no data is shipped to `HDFS`. Am I missing something?

**Comment by saurabh313** [comment]
logs were incomplete because I rechefed only one node, rechefed the whole cluster now and logs are complete
https://gist.github.com/saurabh313/24528fbaa5cae6fd795ec47475d00234

File: cookbooks/bcpc-hadoop/templates/default/flume_log4j.properties.erb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+log4j.appender.file.MaxFileSize = <%= node[:bcpc][:hadoop][:log4j][:max_file_size] %>
+log4j.appender.file.MaxBackupIndex= <%= node[:bcpc][:hadoop][:log4j][:max_backups] %>
```
Comment (cbaenziger) [inline cookbooks/bcpc-hadoop/templates/default/flume_log4j.properties.erb]: Need a space on both sides of the equals sign

**Comment by saurabh313** [comment]
BUG-56300 will fix permissions on datanode logs so flume can read and aggregate them.

**Pull request closed without merging**